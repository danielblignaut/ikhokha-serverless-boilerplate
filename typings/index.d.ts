declare module 'serverless-offline'
declare module 'serverless-webpack'
declare module 'custom-env' 

declare module '*.json' {
	const content: any
	export default content
}

declare module '*.graphql' {
	const content: any
	export default content
}

declare namespace NodeJS {
	interface ProcessEnv {
		PROXY_API_URL: string
		PROXY_API_TOKEN: string
		PROXY_API_TOKEN_KEY: string
		TEST__SERVER_URL: string
		NODE_ENV: 'production' | 'test' | 'development'
		WEBPACK_SERVE: boolean
	}
}