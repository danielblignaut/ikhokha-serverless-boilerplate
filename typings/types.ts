
export interface CustomContext {

}

export interface OperationsToKeep {
	Query: string[]
	Mutation: string[]
	[key: string]: string[]
}

export interface AwsServer {
	
}