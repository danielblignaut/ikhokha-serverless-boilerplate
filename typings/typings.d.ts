declare module '*.graphql' {
	const content: any
	export default content
}

declare module '*.json' {
	const content: any
	export default content
}

declare module 'serverless-webpack' {
	const lib: {
		webpack: {
			isLocal: boolean
		}
		
		entries: any
	}
	export { lib }
}
