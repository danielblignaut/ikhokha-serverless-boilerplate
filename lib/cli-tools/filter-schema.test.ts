import * as typeDefs from '@ikhokha/test/graphql-files/typeDefs'
import filterSchema from '@ikhokha/lib/cli-tools/filter-schema'
import { OperationsToKeep, TypesToKeep } from '@ikhokha/typings/types'
import { makeExecutableSchema } from 'graphql-tools'
import { printSchema, GraphQLSchema } from 'graphql'

describe('filter schema', () => {
	let operationsToKeep: OperationsToKeep
	let typesToKeep: TypesToKeep
	let schema: GraphQLSchema 

	beforeEach(() => {
		operationsToKeep  = {
			Query: ['getAuthorById', 'getPosts'],
			Mutation: ['deletePost', 'editAuthor']
		}
		typesToKeep = ['Post']
		schema = makeExecutableSchema({ typeDefs: typeDefs.example })
	})

	test('give a valid filter and schema', () => {
		let filteredSchema = filterSchema(schema,operationsToKeep, typesToKeep)
		expect(printSchema(filteredSchema)).toMatchSnapshot()
	})

	test('give no query / mutation filters', () => {
		operationsToKeep = {
			Query: [],
			Mutation: []
		}
		let filteredSchema = filterSchema(schema,operationsToKeep, typesToKeep)
		expect(printSchema(filteredSchema)).toMatchSnapshot()
	})

	test('give no type filters', () => {
		typesToKeep = []
		let filteredSchema = filterSchema(schema,operationsToKeep, typesToKeep)
		expect(printSchema(filteredSchema)).toMatchSnapshot()
	})

	test('give one type filters', () => {
		typesToKeep = [
			'Author',
			'Int',
			'String'
		]
		let filteredSchema = filterSchema(schema,operationsToKeep, typesToKeep)
		expect(printSchema(filteredSchema)).toMatchSnapshot()
	})

	test('give invalid query filter', () => {
		operationsToKeep = {
			Query: ['fetchDog'],
			Mutation: []
		}		
		expect(() => filterSchema(schema, operationsToKeep, typesToKeep)).toThrowError()
	})

	test('give invalid type filter', () => {
		typesToKeep = ['NotReal']
		
		expect(() => filterSchema(schema, operationsToKeep, typesToKeep)).toThrowError()
	})

	test('give invalid mutation filter', () => {
		operationsToKeep = {
			Query: [],
			Mutation: ['fetchDog']
		}
		
		expect(() => filterSchema(schema, operationsToKeep, typesToKeep)).toThrowError()
	})

	test('give mutation filters but no mutations on schema', () => {
		operationsToKeep = {
			Query: [],
			Mutation: ['editAuthor']
		}		
		schema = makeExecutableSchema({typeDefs: typeDefs.exampleWithoutQueryAndMutation})
		expect(() => filterSchema(schema, operationsToKeep, typesToKeep)).toThrowError()
	})

	test('give query filters but no mutations on schema', () => {
		operationsToKeep = {
			Query: ['getPosts'],
			Mutation: []
		}
		schema = makeExecutableSchema({typeDefs: typeDefs.exampleWithoutQueryAndMutation})
		expect(() => filterSchema(schema, operationsToKeep, typesToKeep)).toThrowError()
	})

	test('pass invalid filter key', () => {
		operationsToKeep = {
			Query: [],
			Mutation: [],
			randomKey: []
		}
		
		expect(() => filterSchema(schema, operationsToKeep, typesToKeep)).toThrowError()
	})


})