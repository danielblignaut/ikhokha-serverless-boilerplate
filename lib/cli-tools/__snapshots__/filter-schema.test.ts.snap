// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter schema give a valid filter and schema 1`] = `
"type Post
"
`;

exports[`filter schema give no query / mutation filters 1`] = `
"type Post
"
`;

exports[`filter schema give no type filters 1`] = `
"type Author {
  id: Int!
  firstName: String
  lastName: String

  \\"\\"\\"the list of Posts by this author\\"\\"\\"
  posts: [Post]
}

\\"\\"\\"this schema allows the following mutation:\\"\\"\\"
type Mutation {
  deletePost(postId: Int!): Post
  editAuthor(authorId: Int!): Author
}

type Post {
  id: Int!
  title: String
  author: Author
  votes: Int
}

\\"\\"\\"the schema allows the following query:\\"\\"\\"
type Query {
  getPosts: [Post!]!
  getAuthorById(id: Int): Author!
}
"
`;

exports[`filter schema give one type filters 1`] = `
"type Author {
  id: Int!
  firstName: String
  lastName: String
}
"
`;
